### PR TITLE
Add OAuth Authorization

### PR DESCRIPTION
**Pull request by angusmcleod**
@pmusaraj This one is ready now too. Just a note on the storage of OAuth Apps. I decided to use the Plugin Store there because 

- It's a relatively straightforward use case.
- I can forsee a significant variability in the authentication details we need to save if we add other authorization methods in the future, a bit like in the Auth system in Discourse itself.

Other than that, I'll let the code speak for itself.

### CODE REVIEW

File: plugin.rb [hunk 1]
```diff
@@ -555,6 +564,63 @@
...
+          access_token: tokens[domain]
+        }
+      )
+    end
+  end
+
+  add_to_serializer(
+    :user,
+    :activity_pub_authorizations,
+    include_condition: -> { Site.activity_pub_enabled }) do
```
Comment (pmusaraj) [inline plugin.rb:619]:
```suggestion
    include_condition: -> { DiscourseActivityPub.enabled }) do
```

File: app/controllers/discourse_activity_pub/auth_controller.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    before_action :ensure_site_enabled
+
+    def index
+      redirect_to "/ap/auth/authorizations"
+    end
+
+    protected
+
+    def ensure_site_enabled
+      render_auth_error("not_enabled", 403) unless Site.activity_pub_enabled
```
Comment (pmusaraj) [inline app/controllers/discourse_activity_pub/auth_controller.rb:15]:
```suggestion
      render_auth_error("not_enabled", 403) unless DiscourseActivityPub.enabled
```

File: spec/lib/discourse_activity_pub/user_handler_spec.rb [hunk 3]
```diff
@@ -9,6 +9,19 @@
...
     end
 
+    context "when user has authorized the same actor id" do
+      let!(:user) { Fabricate(:user) }
+
+      before do
+        user.activity_pub_save_actor_id("https://external.com", actor.ap_id)
+      end
+
+      it "assocates the user with the actor" do
```
Comment (pmusaraj) [inline spec/lib/discourse_activity_pub/user_handler_spec.rb:19]:
```suggestion
      it "associates the user with the actor" do
```

**Review by pmusaraj — APPROVED** [review]
Apologies for the delay here @angusmcleod, overall this looks great! Left some minor comments, but this is close to being merged, just a couple of minor changes needed and a rebase from main (there are some conflicts). 

Thanks!

**Comment by angusmcleod** [comment]
Hey @pmusaraj just to let you know that I'll be getting to this early next week.

File: spec/requests/discourse_activity_pub/auth/authorization_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+  it { expect(described_class).to be < DiscourseActivityPub::AuthController }
+
+  describe "#destroy" do
+    context "without an actor id" do
+      it "raises an invalid access error" do
+        get "/ap/auth/oauth/authorize"
+        expect(response.status).to eq(403)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (angusmcleod) [reply spec/requests/discourse_activity_pub/auth/authorization_controller_spec.rb]: Good catch, I've fixed that spec: https://github.com/discourse/discourse-activity-pub/pull/24/commits/de2cd7b657d51ca9806ee74193267717991f0e52

**Comment by angusmcleod** [comment]
@pmusaraj I've resolved all the issues, however I can't repro that one failing test locally. Are you able to repro that failure? If not, would you mind re-running the workflow on github?

**Comment by pmusaraj** [comment]
Thanks, no repro locally and CI now's green too @angusmcleod

**Comment by pmusaraj** [comment]
Thanks again @angusmcleod and apologies for the delay in merging, but it's now done!

**Pull request merged**